job_linux.go

```
@@ -0,0 +1,156 @@
+		return nil, fmt.Errorf("%s: %w", path, err)
+	}
+
+	// open cgroup file to jail clone
+	cgroup, err := os.OpenFile(cgroupJob, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", cgroupJob, err)
+	}
+
+	args := append([]string{jailOp, j.Path}, j.Args...)
+	cmd := exec.CommandContext(ctx, j.jailPath, args...)
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags:   syscall.CLONE_NEWPID | syscall.CLONE_NEWNS | syscall.CLONE_NEWNET,
+		Unshareflags: syscall.CLONE_NEWNS,
+		CgroupFD:     int(cgroup.Fd()),
+		UseCgroupFD:  true,
+	}
+	j.cgroup = cgroup
+
+	return cmd, nil
+}
+
+// NewJobReader returns the io.ReadCloser
+func NewJobReader(ctx context.Context, filename string, d Doner) (io.ReadCloser, error) {
+	l, err := os.OpenFile(filename, os.O_RDONLY, 0660)
+	if err != nil {
+		return nil, err
+	}
+	// reader specific notify
+	fd, err := syscall.InotifyInit1(syscall.IN_CLOEXEC)
+	if err != nil {
+		defer l.Close()
+		return nil, os.NewSyscallError("inotify_init", err)
+	}
+	// watch for writes and close event
+	_, err = syscall.InotifyAddWatch(int(fd), l.Name(), syscall.IN_MODIFY|syscall.IN_CLOSE)
+	if err != nil {
+		return nil, os.NewSyscallError("inotify_add_watch", err)
+	}
+	// close file to unblock reads if context is done
+	file := os.NewFile(uintptr(fd), l.Name())
+	go func() {
+		<-ctx.Done()
+		file.Close()
+	}()
+	return &JobReader{doner: d, logs: l, inotify: file}, nil
+}
+
+// Read reads n bytes into buffer and return EOF only when Job stops
+func (r *JobReader) Read(buffer []byte) (n int, err error) {
+	n, err = r.logs.Read(buffer)
+
+	// wait and ignore EOF until stopped
+	if n == 0 && err == io.EOF && !r.doner.Done() {
+
+		// sufficiently size buffer for events
+		b := make([]byte, syscall.SizeofInotifyEvent*syscall.NAME_MAX+1)
+
+		// return EOF if file close by context
+		if n, err = r.inotify.Read(b); errors.Is(err, fs.ErrClosed) {
+			return 0, io.EOF
+		}
+		// keep reading
+		err = ErrReadAgain
```

So if a user of the library reads and receives an ErrReadAgain are they supposed to ignore the error and call Read again? Should they do so immediately? Should they wait first?

That contract seems a bit unidiomatic. Should this just try reading again itself?

---

You're right. That would make the call easier to use and less error prone.